internal/bridge/managedpostgrescluster/managedpostgrescluster_controller.go

```diff
@@ -0,0 +1,552 @@
+		return ctrl.Result{}, nil
+	}
+
+	// Remove SecretInvalid condition if found
+	invalid := meta.FindStatusCondition(managedpostgrescluster.Status.Conditions,
+		v1beta1.ConditionCreating)
+	if invalid != nil && invalid.Status == metav1.ConditionFalse && invalid.Reason == "SecretInvalid" {
+		meta.RemoveStatusCondition(&managedpostgrescluster.Status.Conditions,
+			v1beta1.ConditionCreating)
+	}
+
+	// END SECRET HANDLING
+
+	// If the ManagedPostgresCluster isn't being deleted, add the finalizer
+	if managedpostgrescluster.ObjectMeta.DeletionTimestamp.IsZero() {
+		if !controllerutil.ContainsFinalizer(managedpostgrescluster, finalizer) {
+			controllerutil.AddFinalizer(managedpostgrescluster, finalizer)
+			if err := r.Update(ctx, managedpostgrescluster); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+		// If the ManagedPostgresCluster is being deleted,
+		// handle the deletion, and remove the finalizer
+	} else {
+		if controllerutil.ContainsFinalizer(managedpostgrescluster, finalizer) {
+			log.Info("deleting cluster", "name", managedpostgrescluster.Spec.ClusterName)
+
+			// TODO(managedpostgrescluster): If is_protected, maybe skip this call, but allow the deletion of the K8s object
+			_, deletedAlready, err := r.NewClient().DeleteCluster(ctx, key, managedpostgrescluster.Status.ID)
+			// Requeue if error
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+
+			if !deletedAlready {
+				return ctrl.Result{RequeueAfter: 1 * time.Second}, err
+			}
+
+			// Remove finalizer if deleted already
+			if deletedAlready {
+				log.Info("cluster deleted", "name", managedpostgrescluster.Spec.ClusterName)
+
+				controllerutil.RemoveFinalizer(managedpostgrescluster, finalizer)
+				if err := r.Update(ctx, managedpostgrescluster); err != nil {
+					return ctrl.Result{}, err
+				}
+			}
+		}
+		// Stop reconciliation as the item is being deleted
+		return ctrl.Result{}, nil
+	}
+
+	// Wonder if there's a better way to handle adding/checking/removing statuses
+	// We did something in the upgrade controller
+	// Exit early if we can't create from this K8s object
+	// unless this K8s object has been changed (compare ObservedGeneration)
+	invalid = meta.FindStatusCondition(managedpostgrescluster.Status.Conditions,
+		v1beta1.ConditionCreating)
+	if invalid != nil &&
+		invalid.Status == metav1.ConditionFalse &&
+		invalid.Message == "ClusterInvalid" &&
+		invalid.ObservedGeneration == managedpostgrescluster.GetGeneration() {
+		return ctrl.Result{}, nil
+	}
+
+	// Remove cluster invalid status if found
+	if invalid != nil &&
+		invalid.Status == metav1.ConditionFalse &&
+		invalid.Reason == "ClusterInvalid" {
+		meta.RemoveStatusCondition(&managedpostgrescluster.Status.Conditions,
+			v1beta1.ConditionCreating)
+	}
+
+	// We should only be missing the ID if no create has been issued
+	// or the create was interrupted and we haven't received the ID.
+	if managedpostgrescluster.Status.ID == "" {
+		// START FIND
```

Comment:
Consider annotation gate or some other block to adoption, so that we only allow it if intentional.